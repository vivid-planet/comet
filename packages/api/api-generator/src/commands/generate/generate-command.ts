import { type CrudGeneratorOptions, type CrudSingleGeneratorOptions } from "@comet/cms-api";
import { CLIHelper } from "@mikro-orm/cli";
import { LazyMetadataStorage } from "@nestjs/graphql/dist/schema-builder/storages/lazy-metadata.storage";
import { Command } from "commander";

import { generateCrud } from "./generateCrud/generate-crud";
import { generateCrudSingle } from "./generateCrudSingle/generate-crud-single";
import { writeGeneratedFiles } from "./utils/write-generated-files";

export const generateCommand = new Command("generate").action(async (options) => {
    try {
        const orm = await CLIHelper.getORM(undefined, undefined, { dbName: "generator" });

        const entities = orm.em.getMetadata().getAll();
        LazyMetadataStorage.load();

        for (const name in entities) {
            const entity = entities[name];
            if (!entity.class) {
                // Ignore e.g. relation entities that don't have a class
                continue;
            }
            {
                const generatorOptions = Reflect.getMetadata(`data:crudGeneratorOptions`, entity.class) as CrudGeneratorOptions | undefined;
                if (generatorOptions) {
                    const files = await generateCrud(generatorOptions, entity);
                    await writeGeneratedFiles(files, { targetDirectory: generatorOptions.targetDirectory });
                }
            }
            {
                const generatorOptions = Reflect.getMetadata(`data:crudSingleGeneratorOptions`, entity.class) as
                    | CrudSingleGeneratorOptions
                    | undefined;
                if (generatorOptions) {
                    const files = await generateCrudSingle(generatorOptions, entity);
                    await writeGeneratedFiles(files, { targetDirectory: generatorOptions.targetDirectory });
                }
            }
        }
        await orm.close(true);
    } catch (e) {
        console.warn(e);
    }
});
