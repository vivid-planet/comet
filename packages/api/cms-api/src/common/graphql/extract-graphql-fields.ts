import { GraphQLResolveInfo } from "graphql";
import { parseResolveInfo, ResolveTree } from "graphql-parse-resolve-info";

export function extractGraphqlFields(info: GraphQLResolveInfo, options: { root?: string } = {}): string[] {
    const resolveTree = parseResolveInfo(info) as ResolveTree;
    function treeToList(resolveTree: ResolveTree): string[] {
        const ret = [];
        for (const typeName in resolveTree.fieldsByTypeName) {
            const fields = resolveTree.fieldsByTypeName[typeName];
            for (const fieldName in fields) {
                ret.push(fields[fieldName].name);
                for (const i of treeToList(fields[fieldName])) {
                    ret.push(`${fields[fieldName].name}.${i}`);
                }
            }
        }
        return ret;
    }
    return treeToList(resolveTree).reduce((acc, i) => {
        if (!options.root) return [...acc, i];
        if (!i.startsWith(`${options.root}.`)) {
            return acc;
        }
        return [...acc, i.substring(options.root.length + 1)];
    }, [] as string[]);
}
