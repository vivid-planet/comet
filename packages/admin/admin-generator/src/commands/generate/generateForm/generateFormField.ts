import { type IntrospectionEnumType, type IntrospectionNamedTypeRef, type IntrospectionQuery } from "graphql";

import { type FormConfig, type FormFieldConfig, type GQLDocumentConfigMap } from "../generate-command";
import { camelCaseToHumanReadable } from "../utils/camelCaseToHumanReadable";
import { convertConfigImport } from "../utils/convertConfigImport";
import { type Imports } from "../utils/generateImportsCode";
import { isFieldOptional } from "../utils/isFieldOptional";
import { isGeneratorConfigCode, isGeneratorConfigImport } from "../utils/runtimeTypeGuards";
import { generateAsyncSelect } from "./asyncSelect/generateAsyncSelect";
import { buildFormFieldOptions } from "./formField/options";
import { type GenerateFieldsReturn } from "./generateFields";
import { type Prop } from "./generateForm";

export function generateFormField({
    gqlIntrospection,
    baseOutputFilename,
    config,
    formConfig,
    gqlType,
    namePrefix,
}: {
    gqlIntrospection: IntrospectionQuery;
    baseOutputFilename: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: FormFieldConfig<any>;
    formFragmentName: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    formConfig: FormConfig<any>;
    gqlType: string;
    namePrefix?: string;
}): GenerateFieldsReturn {
    if (config.type == "asyncSelect" || config.type == "asyncSelectFilter") {
        return generateAsyncSelect({ gqlIntrospection, baseOutputFilename, config, formConfig, gqlType, namePrefix });
    }
    const imports: Imports = [];
    const formProps: Prop[] = [];

    const {
        name,
        formattedMessageRootId,
        fieldLabel,
        startAdornment,
        endAdornment,
        imports: optionsImports,
    } = buildFormFieldOptions({ config, formConfig, gqlType, gqlIntrospection });
    imports.push(...optionsImports);

    let { introspectionFieldType } = buildFormFieldOptions({ config, formConfig, gqlType, gqlIntrospection });

    const nameWithPrefix = `${namePrefix ? `${namePrefix}.` : ``}${name}`;

    const rootGqlType = formConfig.gqlType;

    const dataRootName = rootGqlType[0].toLowerCase() + rootGqlType.substring(1); // TODO should probably be deteced via query

    const required = !isFieldOptional({ config, gqlIntrospection, gqlType });

    //TODO verify introspectionField.type is compatbile with config.type

    const endAdornmentWithLockIconProp = `endAdornment={<InputAdornment position="end"><Lock /></InputAdornment>}`;
    const readOnlyProps = `readOnly disabled`;
    const readOnlyPropsWithLock = `${readOnlyProps} ${endAdornmentWithLockIconProp}`;

    let formValuesConfig: GenerateFieldsReturn["formValuesConfig"] = [{}]; // FormFields should only contain one entry

    const gqlDocuments: GQLDocumentConfigMap = {};
    const hooksCode = "";
    let finalFormConfig: GenerateFieldsReturn["finalFormConfig"];

    let validateCode = "";
    if (config.validate) {
        if (isGeneratorConfigImport(config.validate)) {
            imports.push(convertConfigImport(config.validate));
            validateCode = `validate={${config.validate.name}}`;
        } else if (isGeneratorConfigCode(config.validate)) {
            validateCode = `validate={${config.validate.code}}`;
            imports.push(...config.validate.imports.map((imprt) => convertConfigImport(imprt)));
        }
    }

    let code = "";
    let formValueToGqlInputCode = "";
    let formFragmentFields = [name];
    if (config.type == "text") {
        const TextInputComponent = config.multiline ? "TextAreaField" : "TextField";
        code = `
        <${TextInputComponent}
            ${required ? "required" : ""}
            ${config.readOnly ? readOnlyPropsWithLock : ""}
            variant="horizontal"
            fullWidth
            name="${nameWithPrefix}"
            label={${fieldLabel}}
            ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
            ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
            ${
                config.helperText
                    ? `helperText={<FormattedMessage id=` +
                      `"${formattedMessageRootId}.${name}.helperText" ` +
                      `defaultMessage="${config.helperText}" />}`
                    : ""
            }
            ${validateCode}
        />`;
        if (!required && !config.readOnly) {
            formValueToGqlInputCode = `${name}: formValues.${name} ?? null,`;
        }
    } else if (config.type == "number") {
        code = `
            <NumberField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.decimals ? `decimals={${config.decimals}}` : ""}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        //TODO MUI suggest not using type=number https://mui.com/material-ui/react-text-field/#type-quot-number-quot
        let assignment = `parseFloat(formValues.${nameWithPrefix})`;
        if (isFieldOptional({ config, gqlIntrospection: gqlIntrospection, gqlType: gqlType })) {
            assignment = `formValues.${nameWithPrefix} ? ${assignment} : null`;
        }
        formValueToGqlInputCode = `${name}: ${assignment},`;

        let initializationAssignment = `String(data.${dataRootName}.${nameWithPrefix})`;
        if (!required) {
            initializationAssignment = `data.${dataRootName}.${nameWithPrefix} ? ${initializationAssignment} : undefined`;
        }
        formValuesConfig = [
            {
                omitFromFragmentType: name,
                typeCode: `${name}${!required ? `?` : ``}: string;`,
                initializationCode: `${name}: ${initializationAssignment}`,
            },
        ];
    } else if (config.type === "numberRange") {
        code = `
            <Field
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                component={FinalFormRangeInput}
                label={${fieldLabel}}
                min={${config.minValue}}
                max={${config.maxValue}}
                ${config.disableSlider ? "disableSlider" : ""}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;

        formFragmentFields = [`${name}.min`, `${name}.max`];
    } else if (config.type == "boolean") {
        code = `<CheckboxField
                        label={${fieldLabel}}
                        name="${nameWithPrefix}"
                        fullWidth
                        variant="horizontal"
                        ${config.readOnly ? readOnlyProps : ""}
                        ${
                            config.helperText
                                ? `helperText={<FormattedMessage id=` +
                                  `"${formattedMessageRootId}.${name}.helperText" ` +
                                  `defaultMessage="${config.helperText}" />}`
                                : ""
                        }
                        ${validateCode}
                    />`;
        formValuesConfig = [
            {
                defaultInitializationCode: `${name}: false`,
            },
        ];
    } else if (config.type == "date") {
        imports.push({
            name: "Future_DatePickerField",
            importPath: "@comet/admin",
        });
        code = `
            <Future_DatePickerField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        if (!required && !config.readOnly) {
            formValueToGqlInputCode = `${name}: formValues.${name} ?? null,`;
        }
    } else if (config.type == "dateTime") {
        code = `<DateTimeField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        formValuesConfig = [
            {
                initializationCode: `${name}: data.${dataRootName}.${nameWithPrefix} ? new Date(data.${dataRootName}.${nameWithPrefix}) : undefined`,
                omitFromFragmentType: name,
                typeCode: `${name}${!required ? "?" : ""}: Date${!required ? " | null" : ""};`,
            },
        ];
        if (!config.readOnly) {
            formValueToGqlInputCode = required
                ? `${name}: formValues.${name}.toISOString(),`
                : `${name}: formValues.${name} ? formValues.${name}.toISOString() : null,`;
        }
    } else if (config.type == "block") {
        code = `<Field name="${nameWithPrefix}" isEqual={isEqual} label={${fieldLabel}} variant="horizontal" fullWidth>
            {createFinalFormBlock(rootBlocks.${String(config.name)})}
        </Field>`;
        formValueToGqlInputCode = `${name}: rootBlocks.${name}.state2Output(formValues.${nameWithPrefix}),`;
        formValuesConfig = [
            {
                typeCode: `${name}: BlockState<typeof rootBlocks.${name}>;`,
                initializationCode: `${name}: rootBlocks.${name}.input2State(data.${dataRootName}.${nameWithPrefix})`,
                defaultInitializationCode: `${name}: rootBlocks.${name}.defaultValues()`,
            },
        ];
    } else if (config.type === "fileUpload") {
        const multiple = config.multiple || (typeof config.maxFiles === "number" && config.maxFiles > 1);
        code = `<FileUploadField name="${name}" label={${fieldLabel}}
            variant="horizontal"
            ${config.multiple ? "multiple" : ""}
            ${config.maxFiles ? `maxFiles={${config.maxFiles}}` : ""}
            ${config.maxFileSize ? `maxFileSize={${config.maxFileSize}}` : ""}
            ${config.readOnly ? `readOnly` : ""}
            ${config.layout ? `layout="${config.layout}"` : ""}
            ${config.accept ? `accept="${config.accept}"` : ""}
        />`;
        if (multiple) {
            formValueToGqlInputCode = `${name}: formValues.${name}?.map(({ id }) => id),`;
        } else {
            formValueToGqlInputCode = `${name}: formValues.${name} ? formValues.${name}.id : null,`;
        }
        formFragmentFields = [`${name} { ...${config.download ? "FinalFormFileUploadDownloadable" : "FinalFormFileUpload"} }`];
    } else if (config.type == "staticSelect") {
        const multiple = introspectionFieldType?.kind === "LIST";
        if (introspectionFieldType?.kind === "LIST") {
            introspectionFieldType =
                introspectionFieldType.ofType.kind === "NON_NULL" ? introspectionFieldType.ofType.ofType : introspectionFieldType.ofType;
        }

        const enumType = gqlIntrospection.__schema.types.find(
            (t) => t.kind === "ENUM" && t.name === (introspectionFieldType as IntrospectionNamedTypeRef).name,
        ) as IntrospectionEnumType | undefined;
        if (!enumType) throw new Error(`Enum type ${(introspectionFieldType as IntrospectionNamedTypeRef).name} not found for field ${name}`);

        const values = (config.values ? config.values : enumType.enumValues.map((i) => i.name)).map((value) => {
            if (typeof value === "string") {
                return {
                    value,
                    label: camelCaseToHumanReadable(value),
                };
            } else {
                return value;
            }
        });

        let inputType = config.inputType;
        if (!inputType) {
            if (!required || multiple) {
                // radio is not clearable, render always as select if not required
                // radio doesn't support multiple
                inputType = "select";
            } else {
                // auto select/radio based on number of values
                if (values.length <= 5) {
                    inputType = "radio";
                } else {
                    inputType = "select";
                }
            }
        }
        if (inputType === "radio" && multiple) {
            throw new Error(`${name}: inputType=radio doesn't support multiple`);
        }
        if (inputType === "radio" && !required) {
            throw new Error(`${name}: inputType=radio must be required as it doesn't support clearable`);
        }

        const renderAsRadio = inputType === "radio";
        if (renderAsRadio) {
            code = `<RadioGroupField
             ${required ? "required" : ""}
              variant="horizontal"
             fullWidth
             name="${nameWithPrefix}"
             label={${fieldLabel}}
             options={[
                  ${values
                      .map((value) => {
                          return `{
                                label: <FormattedMessage id="${formattedMessageRootId}.${name}.${
                                    value.value.charAt(0).toLowerCase() + value.value.slice(1)
                                }" defaultMessage="${value.label}" />,
                                value: "${value.value}",
                            }`;
                      })
                      .join(",")}
            ]}/>`;
        } else {
            imports.push({
                name: "SelectField",
                importPath: "@comet/admin",
            });
            code = `<SelectField
                            ${required ? "required" : ""}
                             fullWidth
                                variant={"horizontal"}
                                name="${nameWithPrefix}"
                                label={${fieldLabel}}
                                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                                ${
                                    config.helperText
                                        ? `helperText={<FormattedMessage id=` +
                                          `"${formattedMessageRootId}.${name}.helperText" ` +
                                          `defaultMessage="${config.helperText}" />}`
                                        : ""
                                }
                                ${validateCode}
                                ${config.readOnly ? readOnlyPropsWithLock : ""}
                                ${multiple ? "multiple" : ""}
                                options={[${values.map((value) => {
                                    const id = `${formattedMessageRootId}.${name}.${value.value.charAt(0).toLowerCase() + value.value.slice(1)}`;

                                    return `{
                                        value: "${value.value}",
                                        label: <FormattedMessage id="${id}" defaultMessage="${value.label}" />
                                    }`;
                                })}]}
                            />`;
        }
    } else {
        throw new Error(`Unsupported type`);
    }
    return {
        code,
        hooksCode,
        formValueToGqlInputCode,
        formFragmentFields,
        gqlDocuments,
        imports,
        formProps,
        formValuesConfig,
        finalFormConfig,
    };
}
