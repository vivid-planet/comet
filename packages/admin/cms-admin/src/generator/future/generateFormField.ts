import { IntrospectionEnumType, IntrospectionNamedTypeRef, IntrospectionObjectType, IntrospectionQuery } from "graphql";

import { FormConfig, FormFieldConfig, GeneratorReturn } from "./generator";
import { camelCaseToHumanReadable } from "./utils/camelCaseToHumanReadable";
import { Imports } from "./utils/generateImportsCode";

export function generateFormField(
    { gqlIntrospection }: { gqlIntrospection: IntrospectionQuery },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: FormFieldConfig<any>,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    formConfig: FormConfig<any>,
): GeneratorReturn & { imports: Imports } {
    const gqlType = formConfig.gqlType;
    const instanceGqlType = gqlType[0].toLowerCase() + gqlType.substring(1);

    const name = String(config.name);
    const label = config.label ?? camelCaseToHumanReadable(name);

    const introspectionObject = gqlIntrospection.__schema.types.find((type) => type.kind === "OBJECT" && type.name === gqlType) as
        | IntrospectionObjectType
        | undefined;
    if (!introspectionObject) throw new Error(`didn't find object ${gqlType} in gql introspection`);

    const introspectionField = introspectionObject.fields.find((field) => field.name === name);
    if (!introspectionField) throw new Error(`didn't find field ${name} in gql introspection type ${gqlType}`);
    const introspectionFieldType = introspectionField.type.kind === "NON_NULL" ? introspectionField.type.ofType : introspectionField.type;

    const requiredByIntrospection = introspectionField.type.kind == "NON_NULL";

    const required = config.required ?? requiredByIntrospection; //if undefined default to requiredByIntrospection

    //TODO verify introspectionField.type is compatbile with config.type

    const imports: Imports = [];

    let validateCode = "";
    if (config.validate) {
        let importPath = config.validate.import;
        if (importPath.startsWith("./")) {
            //go one level up as generated files are in generated subfolder
            importPath = `.${importPath}`;
        }
        imports.push({
            name: config.validate.name,
            importPath,
        });
        validateCode = `validate={${config.validate.name}}`;
    }

    let code = "";
    if (config.type == "text") {
        code = `
        <Field
            ${required ? "required" : ""}
            ${config.multiline ? "multiline" : ""}
            fullWidth
            name="${name}"
            component={FinalFormInput}
            label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
            ${
                config.helperText
                    ? `helperText={<FormattedMessage id=` + `"${instanceGqlType}.${name}.helperText" ` + `defaultMessage="${config.helperText}" />}`
                    : ""
            }
            ${validateCode}
        />`;
    } else if (config.type == "number") {
        code = `
            <Field
                ${required ? "required" : ""}
                fullWidth
                name="${name}"
                component={FinalFormInput}
                type="number"
                label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${instanceGqlType}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        //TODO MUI suggest not using type=number https://mui.com/material-ui/react-text-field/#type-quot-number-quot
    } else if (config.type == "boolean") {
        code = `<Field name="${name}" label="" type="checkbox" fullWidth ${validateCode}>
            {(props) => (
                <FormControlLabel
                    label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
                    control={<FinalFormCheckbox {...props} />}
                    ${
                        config.helperText
                            ? `helperText={<FormattedMessage id=` +
                              `"${instanceGqlType}.${name}.helperText" ` +
                              `defaultMessage="${config.helperText}" />}`
                            : ""
                    }
                />
            )}
        </Field>`;
    } else if (config.type == "date") {
        code = `
            <Field
                ${required ? "required" : ""}
                fullWidth
                name="${name}"
                component={FinalFormDatePicker}
                label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${instanceGqlType}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
    } else if (config.type == "block") {
        imports.push({
            name: config.block.name,
            importPath: config.block.import,
        });
        code = `<Field name="${name}" isEqual={isEqual}>
            {createFinalFormBlock(${config.block.name})}
        </Field>`;
    } else if (config.type == "staticSelect") {
        if (config.values) {
            throw new Error("custom values for staticSelect is not yet supported"); // TODO add support
        }
        const enumType = gqlIntrospection.__schema.types.find(
            (t) => t.kind === "ENUM" && t.name === (introspectionFieldType as IntrospectionNamedTypeRef).name,
        ) as IntrospectionEnumType | undefined;
        if (!enumType) throw new Error(`Enum type ${(introspectionFieldType as IntrospectionNamedTypeRef).name} not found for field ${name}`);
        const values = enumType.enumValues.map((i) => i.name);
        code = `<Field
            fullWidth
            name="${name}"
            label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}>
            ${
                config.helperText
                    ? `helperText={<FormattedMessage id=` + `"${instanceGqlType}.${name}.helperText" ` + `defaultMessage="${config.helperText}" />}`
                    : ""
            }
            ${validateCode}
            {(props) => 
                <FinalFormSelect {...props}>
                ${values
                    .map((value) => {
                        const id = `${instanceGqlType}.${name}.${value.charAt(0).toLowerCase() + value.slice(1)}`;
                        const label = `<FormattedMessage id="${id}" defaultMessage="${camelCaseToHumanReadable(value)}" />`;
                        return `<MenuItem value="${value}">${label}</MenuItem>`;
                    })
                    .join("\n")}
                </FinalFormSelect>
            }
        </Field>`;
    } else {
        throw new Error(`Unsupported type: ${config.type}`);
    }
    return {
        code,
        gqlDocuments: {},
        imports,
    };
}
