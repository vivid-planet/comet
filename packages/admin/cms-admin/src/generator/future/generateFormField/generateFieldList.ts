import { IntrospectionField, IntrospectionQuery, IntrospectionType } from "graphql";

export function generateFieldListFromIntrospection(
    gqlIntrospection: IntrospectionQuery,
    type: string,
): { path: string; field: IntrospectionField }[] {
    return fieldListFromIntrospectionTypeRecursive(gqlIntrospection.__schema.types, type);
}

function fieldListFromIntrospectionTypeRecursive(
    types: readonly IntrospectionType[],
    type: string,
    parentPath?: string,
): { path: string; field: IntrospectionField }[] {
    const typeDef = types.find((typeDef) => typeDef.name === type);
    if (!typeDef || typeDef.kind !== "OBJECT") return [];

    return typeDef.fields.reduce<{ path: string; field: IntrospectionField }[]>((acc, field) => {
        const path = `${parentPath ? `${parentPath}.` : ""}${field.name}`;
        let type = field.type;
        if (type.kind === "NON_NULL") type = type.ofType;

        if (type.kind === "OBJECT") {
            acc.push({ path: path, field: field });
            acc.push(...fieldListFromIntrospectionTypeRecursive(types, type.name, path));
        } else {
            acc.push({ path: path, field: field });
        }
        return acc;
    }, []);
}
