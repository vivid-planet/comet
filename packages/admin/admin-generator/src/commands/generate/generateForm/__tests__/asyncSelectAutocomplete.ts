import { buildSchema, introspectionFromSchema } from "graphql";

import { type FormConfig, type FormFieldConfig } from "../../generate-command";
import { generateFormField } from "../generateFormField";

describe("AsyncSelect autocomplete", () => {
    it("generates field with default autocomplete because of search arg", async () => {
        const schema = buildSchema(`
            type Query {
                products: [Product]!
                categories(search: String): [Category]!
            } 

            type Product {
                id: ID!
                title: String!
                category: Category!
            }
            type Category {
                id: ID!
                name: String!
            }

            type Mutation {
                createProduct(input: ProductInput!): Product!
                updateProduct(id: ID!, input: ProductInput!): Product!
            }

            input ProductInput {
                title: String
                category: ID
            }
        `);
        type GQLCategory = {
            __typename?: "Category";
            id: string;
            name: string;
        };
        type GQLProduct = {
            __typename?: "Product";
            id: string;
            title: string;
            category: GQLCategory;
        };

        const introspection = introspectionFromSchema(schema);

        const fieldConfig: FormFieldConfig<GQLProduct> = {
            type: "asyncSelect",
            rootQuery: "categories",
            name: "category",
        };
        const formConfig: FormConfig<GQLProduct> = {
            type: "form",
            gqlType: "Product",
            fields: [fieldConfig],
        };

        const output = generateFormField({
            gqlIntrospection: introspection,
            baseOutputFilename: "ProductForm",
            formFragmentName: "ProductFormFragment",
            config: fieldConfig,
            formConfig,
            gqlType: "Product",
        });

        expect(output.code).toMatchSnapshot();
    });

    it("generates field without autocomplete explicitly set", async () => {
        const schema = buildSchema(`
            type Query {
                products: [Product]!
                categories(search: String): [Category]!
            } 

            type Product {
                id: ID!
                title: String!
                category: Category!
            }
            type Category {
                id: ID!
                name: String!
            }

            type Mutation {
                createProduct(input: ProductInput!): Product!
                updateProduct(id: ID!, input: ProductInput!): Product!
            }

            input ProductInput {
                title: String
                category: ID
            }
        `);
        type GQLCategory = {
            __typename?: "Category";
            id: string;
            name: string;
        };
        type GQLProduct = {
            __typename?: "Product";
            id: string;
            title: string;
            category: GQLCategory;
        };

        const introspection = introspectionFromSchema(schema);

        const fieldConfig: FormFieldConfig<GQLProduct> = {
            type: "asyncSelect",
            rootQuery: "categories",
            name: "category",
            autocomplete: false,
        };
        const formConfig: FormConfig<GQLProduct> = {
            type: "form",
            gqlType: "Product",
            fields: [fieldConfig],
        };

        const output = generateFormField({
            gqlIntrospection: introspection,
            baseOutputFilename: "ProductForm",
            formFragmentName: "ProductFormFragment",
            config: fieldConfig,
            formConfig,
            gqlType: "Product",
        });

        expect(output.code).toMatchSnapshot();
    });

    it("throws error if autocomplete is true but there is no search arg", async () => {
        const schema = buildSchema(`
            type Query {
                products: [Product]!
                categories: [Category]!
            } 

            type Product {
                id: ID!
                title: String!
                category: Category!
            }
            type Category {
                id: ID!
                name: String!
            }

            type Mutation {
                createProduct(input: ProductInput!): Product!
                updateProduct(id: ID!, input: ProductInput!): Product!
            }

            input ProductInput {
                title: String
                category: ID
            }
        `);
        type GQLCategory = {
            __typename?: "Category";
            id: string;
            name: string;
        };
        type GQLProduct = {
            __typename?: "Product";
            id: string;
            title: string;
            category: GQLCategory;
        };

        const introspection = introspectionFromSchema(schema);

        const fieldConfig: FormFieldConfig<GQLProduct> = {
            type: "asyncSelect",
            rootQuery: "categories",
            name: "category",
            autocomplete: true,
        };
        const formConfig: FormConfig<GQLProduct> = {
            type: "form",
            gqlType: "Product",
            fields: [fieldConfig],
        };

        expect(() => {
            generateFormField({
                gqlIntrospection: introspection,
                baseOutputFilename: "ProductForm",
                formFragmentName: "ProductFormFragment",
                config: fieldConfig,
                formConfig,
                gqlType: "Product",
            });
        }).toThrow();
    });

    it("throws error if search arg is no String", async () => {
        const schema = buildSchema(`
            type Query {
                products: [Product]!
                categories(search: Int): [Category]!
            } 

            type Product {
                id: ID!
                title: String!
                category: Category!
            }
            type Category {
                id: ID!
                name: String!
            }

            type Mutation {
                createProduct(input: ProductInput!): Product!
                updateProduct(id: ID!, input: ProductInput!): Product!
            }

            input ProductInput {
                title: String
                category: ID
            }
        `);
        type GQLCategory = {
            __typename?: "Category";
            id: string;
            name: string;
        };
        type GQLProduct = {
            __typename?: "Product";
            id: string;
            title: string;
            category: GQLCategory;
        };

        const introspection = introspectionFromSchema(schema);

        const fieldConfig: FormFieldConfig<GQLProduct> = {
            type: "asyncSelect",
            rootQuery: "categories",
            name: "category",
        };
        const formConfig: FormConfig<GQLProduct> = {
            type: "form",
            gqlType: "Product",
            fields: [fieldConfig],
        };

        expect(() => {
            generateFormField({
                gqlIntrospection: introspection,
                baseOutputFilename: "ProductForm",
                formFragmentName: "ProductFormFragment",
                config: fieldConfig,
                formConfig,
                gqlType: "Product",
            });
        }).toThrow();
    });
});
