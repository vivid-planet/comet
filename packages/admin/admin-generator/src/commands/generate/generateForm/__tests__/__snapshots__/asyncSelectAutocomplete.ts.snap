// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncSelect autocomplete generates field with default autocomplete because of search arg 1`] = `
"<AsyncAutocompleteField
                required
                variant="horizontal"
                fullWidth
                
                name="category"
                label={<FormattedMessage id="product.category" defaultMessage="Category" />}
                
                loadOptions={async (search?: string) => {
                    const { data } = await client.query<GQLCategoriesSelectQuery, GQLCategoriesSelectQueryVariables>({
                        query: gql\`query CategoriesSelect(
                            
                            
                            $search: String
                        ) {
                            categories(
                                
                                search: $search
                            ) {
                                nodes {
                                    id
                                    name
                                }
                            }
                        }\`, variables: { 
                            
                            search,
                         }
                    });
                    return data.categories.nodes;
                }}
                getOptionLabel={(option) => option.name}
                
            />"
`;

exports[`AsyncSelect autocomplete generates field without autocomplete explicitly set 1`] = `
"<AsyncSelectField
                required
                variant="horizontal"
                fullWidth
                
                name="category"
                label={<FormattedMessage id="product.category" defaultMessage="Category" />}
                
                loadOptions={async () => {
                    const { data } = await client.query<GQLCategoriesSelectQuery, GQLCategoriesSelectQueryVariables>({
                        query: gql\`query CategoriesSelect
                            
                            
                            
                         {
                            categories
                                
                                
                             {
                                nodes {
                                    id
                                    name
                                }
                            }
                        }\`
                            
                            
                        
                    });
                    return data.categories.nodes;
                }}
                getOptionLabel={(option) => option.name}
                
            />"
`;
